tasks/ast_codegen/src/generators/derive_clone_in.rs

```diff
@@ -0,0 +1,134 @@
+use itertools::Itertools;
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::Ident;
+
+use crate::{
+    output,
+    schema::{EnumDef, GetIdent, StructDef, TypeDef},
+    GeneratorOutput, LateCtx,
+};
+
+use super::{define_generator, generated_header, Generator};
+
+define_generator! {
+    pub struct DeriveCloneIn;
+}
+
+impl Generator for DeriveCloneIn {
+    fn name(&self) -> &'static str {
+        stringify!(DeriveCloneIn)
+    }
+
+    fn generate(&mut self, ctx: &LateCtx) -> GeneratorOutput {
+        let impls: Vec<TokenStream> = ctx
+            .schema
+            .definitions
+            .iter()
+            .filter(|def| def.generates_derive("CloneIn"))
+            .map(|def| match &def {
+                TypeDef::Enum(it) => derive_enum(it),
+                TypeDef::Struct(it) => derive_struct(it),
+            })
+            .collect();
+
+        let header = generated_header!();
+
+        GeneratorOutput::Stream((
+            output(crate::AST_CRATE, "derive_clone_in.rs"),
+            quote! {
+                #header
+
+                use oxc_allocator::{Allocator, CloneIn};
+                use crate::ast::*;
+                endl!();
+
+                #(#impls)*
+            },
+        ))
+    }
+}
+
+fn derive_enum(def: &EnumDef) -> TokenStream {
+    let ty_ident = def.ident();
+    impl_clone_in(&ty_ident, def.has_lifetime, || {
+        let mut used_alloc = false;
+        let matches = def
+            .all_variants()
+            .map(|var| {
+                let ident = var.ident();
+                if var.is_unit() {
+                    quote!(Self :: #ident => Self :: Cloned :: #ident)
+                } else {
+                    used_alloc = true;
+                    quote!(Self :: #ident(it) => Self :: Cloned :: #ident(it.clone_in(alloc)))
+                }
+            })
+            .collect_vec();
+        let alloc_ident = if used_alloc { format_ident!("alloc") } else { format_ident!("_") };
+        (
+            alloc_ident,
+            quote! {
+                match self {
+                    #(#matches),*
+                }
+            },
+        )
+    })
+}
+
+fn derive_struct(def: &StructDef) -> TokenStream {
+    let ty_ident = def.ident();
+    impl_clone_in(&ty_ident, def.has_lifetime, || {
+        let (alloc_ident, body) = if def.fields.is_empty() {
+            (format_ident!("_"), TokenStream::default())
+        } else {
+            let fields = def.fields.iter().map(|field| {
+                let ident = field.ident();
+                quote!(#ident: self.#ident.clone_in(alloc))
+            });
+            (format_ident!("alloc"), quote!({ #(#fields),* }))
+        };
+        (
+            alloc_ident,
+            quote! {
+                    #ty_ident #body
```

Comment:
I think this can just be `Self #body`? I don't think it matters if the lifetimes (`'old_alloc` vs `'new_alloc`) don't match here.

Reply:
The lifetime would be a problem here, But if you don't like writing the actual identifier here I can use `Self::Cloned` instead.

Reply:
I *think* `Self` would work, but maybe I'm wrong. I'll try it once this is merged.

Reply:
I started with `Self`, When it didn't work I resorted to using the identifier. But there might be something I'm missing here. Please go ahead and give it a shot.